pkg/providers/openai/chat.go

```diff
@@ -4,232 +4,147 @@ import (
+		MaxTokens:        cfg.DefaultParams.MaxTokens,
+		N:                cfg.DefaultParams.N,
+		StopWords:        cfg.DefaultParams.StopWords,
+		Stream:           false, // unsupported right now
+		FrequencyPenalty: cfg.DefaultParams.FrequencyPenalty,
+		PresencePenalty:  cfg.DefaultParams.PresencePenalty,
+		LogitBias:        cfg.DefaultParams.LogitBias,
+		User:             cfg.DefaultParams.User,
+		Seed:             cfg.DefaultParams.Seed,
+		Tools:            cfg.DefaultParams.Tools,
+		ToolChoice:       cfg.DefaultParams.ToolChoice,
+		ResponseFormat:   cfg.DefaultParams.ResponseFormat,
 	}
+}
 
-	chatRequestValue := reflect.ValueOf(chatRequest).Elem()
-	chatRequestType := chatRequestValue.Type()
-
-	for i := 0; i < chatRequestValue.NumField(); i++ {
-		jsonTags := strings.Split(chatRequestType.Field(i).Tag.Get("json"), ",")
-		jsonTag := jsonTags[0]
+func NewChatMessagesFromUnifiedRequest(request *schemas.UnifiedChatRequest) []ChatMessage {
+	messages := make([]ChatMessage, 0, len(request.MessageHistory)+1)
 
-		if value, ok := defaultParams[jsonTag]; ok {
-			fieldValue := chatRequestValue.Field(i)
-			fieldValue.Set(reflect.ValueOf(value))
-		}
+	// Add items from messageHistory first and the new chat message last
+	for _, message := range request.MessageHistory {
+		messages = append(messages, ChatMessage{Role: message.Role, Content: message.Content})
 	}
 
-	// c.Telemetry.Logger.Info("chatRequest created", zap.Any("chatRequest body", chatRequest))
+	messages = append(messages, ChatMessage{Role: request.Message.Role, Content: request.Message.Content})
 
-	return chatRequest
+	return messages
 }
 
-// CreateChatResponse creates chat Response.
-func (c *ProviderClient) CreateChatResponse(ctx context.Context, r *ChatRequest, u *providers.UnifiedAPIData) (*ChatResponse, error) {
-	_ = ctx // keep this for future use
+// Chat sends a chat request to the specified OpenAI model.
+func (c *Client) Chat(ctx context.Context, request *schemas.UnifiedChatRequest) (*schemas.UnifiedChatResponse, error) {
+	// Create a new chat request
+	chatRequest := c.createChatRequestSchema(request)
 
-	resp, err := c.createChatHTTP(r, u)
+	// TODO: this is suspicious we do zero remapping of OpenAI response and send it back as is.
+	//  Does it really work well across providers?
+	chatResponse, err := c.doChatRequest(ctx, chatRequest)
 	if err != nil {
 		return nil, err
 	}
 
-	if len(resp.Choices) == 0 {
+	if len(chatResponse.Choices) == 0 {
 		return nil, ErrEmptyResponse
 	}
 
-	return resp, nil
+	return chatResponse, nil
 }
 
-func (c *ProviderClient) createChatHTTP(payload *ChatRequest, u *providers.UnifiedAPIData) (*ChatResponse, error) {
-	c.Telemetry.Logger.Info("running createChatHttp")
+func (c *Client) createChatRequestSchema(request *schemas.UnifiedChatRequest) *ChatRequest {
+	// TODO: consider using objectpool to optimize memory allocation
+	chatRequest := c.chatRequestTemplate // hoping to get a copy of the template
```

Comment:
What do you have in mind for the template? Is it different from the ChatRequest type above?

Reply:
I have noticed that most of the chat request object values are well-defined in advance on the client initialization ('cause they are passed via configs and nothing changes them afterwards as of now). So I have filled a instance of chat request on the client initialization to just copy it further and change fields that vary like messages.

In any case, it's better to benchmark this code to see if this brings any value (Go supports benchmarking as a special type of tests out of the box which is nit).